src/plugins/intel_cpu/src/nodes/mkldnn_non_zero.cpp

```diff
@@ -108,32 +115,123 @@ void MKLDNNNonZeroNode::executeSpecified() {
+    }
 
     if (isDynamicNode()) {
-        VectorDims newDims{inRank, nonZeroCount};
+        VectorDims newDims{inRank, totalNonZeroCount};
         redefineOutputMemory({newDims});
     }
     int *dst = reinterpret_cast<int *>(dstMemPtr->GetPtr());
-    size_t inSize = inShape.getElementsCount();
-    auto srcStrides = getParentEdgeAt(0)->getMemory().GetDescWithType<BlockedMemoryDesc>()->getStrides();
-    if (nonZeroCount == 0)
+    auto srcDims = inShape.getDims();
+    if (totalNonZeroCount == 0)
         return;
-    if (inShape.getRank() == 0) {
+
+    switch (inRank) {
+    case 0:
         dst[0] = 0;
-    } else {
-        size_t colIndex = 0, outIndex = 0;
-        for (size_t i = 0; i < inSize; i++) {
-            if (src[i] != zero) {
-                size_t temp = i;
-                for (size_t j = 0; j < inRank; j++) {
-                    outIndex = j * nonZeroCount + colIndex;
-                    dst[outIndex] = temp / srcStrides[j];
-                    temp = temp % srcStrides[j];
-                }
-                colIndex++;
-            }
-        }
+        break;
+    case 1:
+        parallel_for(srcDims[0],
+                     [&](size_t ithr, size_t i0) {
+                         size_t inputIndex = i0;
+                         size_t& outputIndex = destIndices[ithr];
+
+                         if (src[inputIndex] != zero) {
+                             dst[outputIndex] = static_cast<int>(i0);
+                             outputIndex++;
+                         }
+                     });
+        break;
+    case 2:
+    {
+        size_t i0Stride = srcDims[1];
+
+        parallel_for2d(srcDims[0], srcDims[1],
+                       [&](size_t ithr, size_t i0, size_t i1) {
+                           size_t inputIndex = i0 * i0Stride + i1;
+                           size_t& outputIndex = destIndices[ithr];
+
+                           if (src[inputIndex] != zero) {
+                               dst[outputIndex] = static_cast<int>(i0);
+                               dst[outputIndex + totalNonZeroCount] = static_cast<int>(i1);
+                               outputIndex++;
+                           }
+                       });
+        break;
+    }
+    case 3:
+    {
+        size_t i1Stride = srcDims[2];
+        size_t i0Stride = srcDims[1] * i1Stride;
+
+        parallel_for3d(srcDims[0], srcDims[1], srcDims[2],
+                       [&](size_t ithr, size_t i0, size_t i1, size_t i2) {
+                           size_t inputIndex = i0 * i0Stride + i1 * i1Stride + i2;
+                           size_t& outputIndex = destIndices[ithr];
+
+                           if (src[inputIndex] != zero) {
+                               dst[outputIndex] = static_cast<int>(i0);
+                               dst[outputIndex + totalNonZeroCount] = static_cast<int>(i1);
+                               dst[outputIndex + totalNonZeroCount * 2] = static_cast<int>(i2);
```

Comment:
Here and bellow: There are multiplications which could be moved out of the loop:
```C++
        auto 2xTotalNonZeroCount = 2 * totalNonZeroCount;
        parallel_for3d(srcDims[0], srcDims[1], srcDims[2],
                       [&](size_t ithr, size_t i0, size_t i1, size_t i2) {
                           // ...
                           dst[outputIndex + 2xTotalNonZeroCount] = static_cast<int>(i2);
                           // ...